lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,130 @@
+
+
+use Omnipay\Omnipay as OPay;
+use Aimeos\MShop\Order\Item\Base as Status;
+
+use Illuminate\Support\Facades\Auth;
+
+
+/**
+ * Payment provider for payment gateways supported by the PaypalPlus library.
+ *
+ * @package MShop
+ * @subpackage Service
+ */
+class PaypalPlus
+extends    \Aimeos\MShop\Service\Provider\Payment\OmniPay
+//	extends \Aimeos\MShop\Service\Provider\Payment\Base
+	implements \Aimeos\MShop\Service\Provider\Payment\Iface
+{
+
+	protected function processOrder( \Aimeos\MShop\Order\Item\Iface $order,
+		array $params = [] ) : ?\Aimeos\MShop\Common\Helper\Form\Iface
+	{
+		$parts = \Aimeos\MShop\Order\Item\Base\Base::PARTS_SERVICE
+			| \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT
+			| \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS;
+
+		$base = $this->getOrderBase( $order->getBaseId(), $parts );
+		$data = $this->getData( $base, $order->getId(), $params );
+		$urls = $this->getPaymentUrls();
+
+		$data['cancelUrl'] = $this->getConfigValue( 'cancelUrl', '' ) ;
+
+		if(empty($data['cancelUrl'])) {
+			$data['cancelUrl'] = $_SERVER['HTTP_ORIGIN'] ;
+		}
+		
+		try
+		{
+			$response = $this->sendRequest( $order, $data );
+			$isSuccessful = $response->isSuccessful()  ;
+
+			$approval_url = "" ;
+			$testmode = "" ;
+			$countryid = "" ;
+
+			if( $isSuccessful )
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				$this->saveRepayData( $response, $base->getCustomerId() );
+
+				$status = $this->getValue( 'authorize', false ) ? Status::PAY_AUTHORIZED : Status::PAY_RECEIVED;
+				$this->saveOrder( $order->setPaymentStatus( $status ) );
+				
+				if(!empty($response->getData()['links']['1']['href'])){
+					$approval_url = $response->getData()['links']['1']['href'] ;
+				}
+		
+				
+				$testmode = $this->getConfigValue( 'testmode', '' ) ;
+				if(!empty($testmode) && $testmode=="1"){
+					$testmode = "sandbox" ;
+				}elseif(!empty($testmode) && $testmode=="0"){
+					$testmode = "live" ;
+				}
+	
+				$countryid = Auth::user()->countryid;
+
+
+				if(empty($approval_url)  || empty($testmode)  || empty($countryid) || empty($data['cancelUrl']) ){
+					throw new \Aimeos\MShop\Service\Exception( $response->getMessage() );
+				}
+
+			}
+			elseif( $response->isRedirect() )
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				return $this->getRedirectForm( $response );
+			}
+			else
+			{
+				$this->saveOrder( $order->setPaymentStatus( Status::PAY_REFUSED ) );
+				throw new \Aimeos\MShop\Service\Exception( $response->getMessage() );
+			}
+		}
+		catch( \Exception $e )
+		{
+			throw new \Aimeos\MShop\Service\Exception( $e->getMessage() );
+		} 
+
+
+		
+		return new \Aimeos\MShop\Common\Helper\Form\Standard( $urls['returnUrl'] ?? '', 'POST', [], false ,$this->getPayPalPlusJs($approval_url,$testmode ,$countryid ) );
+
+	}
+
+	protected function getPayPalPlusJs($approval_url,$testmode ,$countryid ) 
+	{
+		echo '
```

Comment:
The `echo` is wrong, you need to `return` the string

Reply:
here is the problem that I have described previously in PR 27 .. 
the issue at js ppp library from Paypal as it demands to be loaded as landing page -unlike of stripe js library- ,so I should terminate the code to force it to stay till ppp is loaded completely .. I just replace die() with exit() .
 now I am spending my time to learn how to create new controller and its view -but unfortunately I don't get any resources about that- as the PaypalPlus is an part of new created extension

Reply:
Doesn't the PayPal code embed into the existing page? Because it's JS code, it should be the case.
Do you have a link to the documentation?

Don't start creating a new controller because it's the wrong way.

Reply:
> Doesn't the PayPal code embed into the existing page? Because it's JS code, it should be the case.
> Do you have a link to the documentation?
yes sure .. here it is : https://developer.paypal.com/docs/integration/paypal-plus/mexico-brazil/integrate-a-payment-selection-page/#display-mini-browser
further more , I have created a pure php and follow instructions and it worked fine ! so the problem raised when we use Aimeos ..
> Don't start creating a new controller because it's the wrong way.
I have started already walk thought as I want to use another option in Paypal Plus Js library which is "use popup window" , but I crashed with a lot of Aimeos coding way to be implemented , for that reason I have create a topic in Aimeos Help "https://aimeos.org/help/help-f15/new-independent-completely-extension-t3346.html" which it is a must for me if I will continue working on Aimeos , no matter if that required for this PR or not .. thanks in advance

Reply:
The docs say that you can add a button to open another browser window or use and iframe. Did you try those options?
Your method can contains arbitrary HTML and/or JS code so both should be possible.

Reply:
yes .. I tried .. 
but same problem .. which is the code doesn't wait ppp js library to be loaded .. it goes to next statements unless I exit() or die() the execution !
please help !

Reply:
Use "onsite: yes" in the admin backend configuration for the payment option to stay at the page. Then, no automatically executed post by the Javascript should occur.

Reply:
I tried this option too and it fails as there are no way to inject required ppp js inside onsite view file  , it just show a credit card and expiration date and other input boxes which I don't need them ... 
but I will try it again

Reply:
Your JS is always injected:
https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L138

You can overwrite this method in your class and return an empty array to avoid rendering any fields:
https://github.com/aimeoscom/ai-payments/blob/master/lib/custom/src/MShop/Service/Provider/Payment/Stripe.php#L117-L126

Reply:
your idea was great , but if there are no fileds , it won't stay at that page and goes through confirmation page without waiting js to be loaded .. so, I have added one filed -firstname for example- and that worked !
but I need to handle this field and need to remove "pay now" button ... 
![aa](https://user-images.githubusercontent.com/19593217/102616192-3a824a00-4148-11eb-8c2e-a61d730616f0.jpeg)

Reply:
Use a hidden field (public => false) to avoid the field.

We can also extend the template and add here: https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L167
```
<?php elseif( $this->get( 'standardMethod', 'POST' ) === '' ) : ?>
	<a class="btn btn-default btn-lg" href="<?= $enc->attr( $this->standardUrlPayment ); ?>">
		<?= $enc->html( $this->translate( 'client', 'Change payment' ), $enc::TRUST ); ?>
	</a>
<?php else : ?>
	<button class="btn btn-primary btn-lg btn-action">
		<?= $enc->html( $this->translate( 'client', 'Proceed' ), $enc::TRUST ); ?>
	</button>
<?php endif; ?>

Reply:
I tried it but unfortunately public => false is same as deleting it .. it makes execution keep run without wait .. 
anyway I will try working this point
thanks a lot

Reply:
If we change the template too, the JS won't POST the form any more. Just use
```
return new \Aimeos\MShop\Common\Helper\Form\Standard( '', '', true ,$this->getPayPalPlusJs($approval_url,$testmode ,$countryid ) );
```
in your payment provider.

Reply:
head shut , it worked :D